### DISCUSSION TITLE
"postgres" package should go away, as it can't auto-upgrade the data and there is no backwards-compat

### DISCUSSION THREAD

**Discussion by mathieulb** [Homebrew/discussions#2285] [Everyday usage] [answered]
brew's upgrade of postgres just broke, as it tried to run 14.* binaries on 13.* data, which makes no sense. If postgresql can't keep any backwards-compatibility in their data formats, and if brew can't provide an upgrade script for this, then major-version-locked packages (@), such as postgresql@13 and postgresql@14, should be all there is to install. Upgrading a formula named postgresql@13 will never make it turn into a broken postgresql@14 overnight.

**Comment by SMillerDev** [comment]
We have a script and installing or upgrading postgres will print a caveat telling you how to use it.

**Reply by carlocab** [reply]
You should also be able to `brew install postgresql@14` and stay on version 14 even when the `postgresql` formula has been updated to version 15.

**Reply by mathieulb** [reply]
SMillerDev, ok, thanks, but I missed it, as I didn't think of scrolling back. Does the log get recorded in any file ? I can't find such a log. There is one when using apt and/or dpkg, iirc. And still, I'd expect that if an upgrade breaks all the data compatibility and even prevents the server from being restarted, and if you can't make it a fully automatic upgrade, there should at least be a prompt about it during upgrade, perhaps even an offer to pin to version 13. I don't know why Github calls your comment a suggested answer though, because my main point is about removing the `postgresql` package.

**Reply by mathieulb** [reply]
carlocab, I already know that, and I'm already mentioning the existence of `postgresql@14` above, my main point is that there might be no reason to keep the package `postgresql` that doesn't have a `@` in the name.

**Reply by carlocab** [reply]
Perhaps you did know it, but that wasn't apparent from your saying

> brew's upgrade of postgres just broke, as it tried to run 14.* binaries on 13.* data, which makes no sense.

In any case, even if you did know it, you're not the only one reading this, and it seems likely that not every person reading this knows that doing `brew install foo@x` will keep you at version `x` as long as `x` continues to exist in Homebrew/core, even if `foo@x` is just an alias for `foo`.

> my main point is that there might be no reason to keep the package `postgresql` that doesn't have a `@` in the name.

Here's one reason. Homebrew offers you a simple contract: `brew install foo` installs the latest version of `foo` available in Homebrew/core. If you don't like that, and there are versioned formulae available, you should `brew install foo@x`. Creating exceptions for this is unnecessarily confusing for users.

**Reply by mathieulb** [reply]
carlocab, I'm suggesting that the lack of exceptions for this is unnecessarily confusing for users, as packages that break like this when upgrading are already an exception to the expectation that any upgrading keeps the package able to run, at the very least if still using the default config file; and then if any other package broke because of an outdated config file, I'd be able to just fix the config file, such as copying the one from the latest package, and that'd be enough to make it run again, but no, postgres(ql) isn't like that.

**Reply by carlocab** [reply]
What should happen when a user does `brew install postgresql`, then?

**Reply by SMillerDev** [reply]
I think the better audience for your complaint is the postgres developers. Brew has their standards about versioning and our postgres formula follows it. We even have the `postgresql-upgrade-database` script to make it easier and we warn people to use it.

I agree postgres shouldn't break on every update, but that's not on brew to fix.

**Reply by mathieulb** [reply]
carlocab, ideally, it would reject it. For example, right now, if I try installing a non-existent package that begins with the same letters, `brew install postgr` says `Warning: No available formula or cask with the name "postgr". Did you mean poster or postgis?
==> Searching for similarly named formulae...
These similarly named formulae were found:
check_postgres   postgresql ✔     postgresql@10    postgresql@11    postgresql@12    postgresql@13 ✔  postgresql@9.4   postgresql@9.5   postgresql@9.6   postgrest        qt-postgresql    poster           postgis`

So, if `postgresql` didn't exist from the start, someone trying to install it anyway using that name would get roughly the same list of possibilities as is listed above. If not, `brew search` would.

But because the package already exists, I don't know what can happen to make it disappear. If it implies auto-deleting the package whenever a user upgrades, the current situation is a lesser evil. If it implies just stopping upgrades with a "warning: package no longer exists in repo", then it might be ok to delete the package, but it depends on taste, and it would be better to postpone this as much as possible (which means just before `postgresql@15` gets added). But if an upgrade procedure can name a successor package name that replaces it, then `postgresql` could name `postgresql@14` as its successor immediately, so that anyone upgrading `postgresql` would be switched to `postgresql@14` as if they had installed `postgresql@14`, instead  of what the current alias system does. This would have to be a supplementary mechanism, as most packages that are aliases for a `@` package don't have this problem with upgrades.

**Reply by mathieulb** [reply]
SMillerDev, I think that the reason this problem exists in brew (after that many years) is because brew developers assume that it's a problem in postgres whereas postgres developers surely would claim it's a problem with brew if we contacted them. But in practice, the problem is passed down to the user who then gets into trouble for having installed `postgresql` instead of a `@` package and not noticing the upgrade message that doesn't get recorded anywhere.

**Reply by SMillerDev** [reply]
It's actually documented as the standard procedure for a postgres update: https://www.postgresql.org/docs/12/upgrading.html

Brew is not doing anything special here, this is standard postgres behavior.

**Reply by mathieulb** [reply]
SMillerDev, what I'm saying is that brew isn't doing anything special and that it should be doing something special so as to adjust to "standard postgres behavior".

**Reply by tlk** [reply]
In https://github.com/Homebrew/discussions/discussions/57#discussioncomment-1429292 a Homebrew user ran `brew install nmap` and later found that this operation had left the previously running and functional postgresql service in a non-functional state.

Chances are that the console output from the `brew install nmap` operation was gone when the user eventually realized that the postgresql service was no longer functional. If the user had not been upgrading/installing Homebrew packages for a while, the console output would likely be massive, due to the many updates, so even if the console output was still available chances are that the crucial information about having to run `brew postgresql-upgrade-database` was lost in the terminal buffer.

In either case, I believe that some (most?)  Homebrew users are surprised to learn that Homebrew will upgrade postgresql (and possibly leave it in a non-functional state) as a side-effect of installing `nmap`.

Given that the default Homebrew policy is to upgrade a database service as a side-effect without asking the end user, perhaps Homebrew could be helpful and run the `brew postgresql-upgrade-database` command automatically when it is upgrading postgresql?

**Comment by carlocab** [comment, marked as answer]
For the record, `postgresql` has gone away. Doing that still broke things, though: #3579, #3584, #3595, Homebrew/homebrew-core#109164

**Reply by mathieulb** [reply]
None of the 4 problems you are pointing to seem to me like they are the result of the removal of the `postgresql` package. The only purpose of the removal was to prevent surprising automatic upgrades that don't convert the computer's pg tables to the new format.
